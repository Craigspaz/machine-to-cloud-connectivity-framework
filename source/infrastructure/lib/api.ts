// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, Construct, RemovalPolicy } from '@aws-cdk/core';
import { ServicePrincipal } from '@aws-cdk/aws-iam';
import { Function as LambdaFunction } from '@aws-cdk/aws-lambda';
import {
  AccessLogFormat,
  AuthorizationType,
  ContentHandling,
  Deployment,
  EndpointType,
  Integration,
  IntegrationType,
  LogGroupLogDestination,
  MethodLoggingLevel,
  MethodOptions,
  PassthroughBehavior,
  RequestValidator,
  RestApi,
  Stage
} from '@aws-cdk/aws-apigateway';
import { LogGroup, RetentionDays } from '@aws-cdk/aws-logs';
import { addCfnSuppressRules } from '../utils/utils';

/**
 * ApiConstructProps props
 * @interface ApiConstructProps
 */
export interface ApiConstructProps {
  // Connection builder Lambda function
  readonly connectionBuilderLambdaFunction: LambdaFunction;
}

/**
 * @class
 * Machine to Cloud Connectivity Framework API Construct.
 * It creates an API Gateway REST API and other resources.
 */
export class ApiConstruct extends Construct {
  // API endpoint
  public apiEndpoint: string;
  // API ID
  public apiId: string;

  constructor(scope: Construct, id: string, props: ApiConstructProps) {
    super(scope, id);

    const { connectionBuilderLambdaFunction } = props;
    const connectionBuilderLambdaFunctionArn = connectionBuilderLambdaFunction.functionArn;

    const apiLogGroup = new LogGroup(this, 'Logs', {
      removalPolicy: RemovalPolicy.DESTROY,
      retention: RetentionDays.THREE_MONTHS
    });
    addCfnSuppressRules(apiLogGroup, [{
      id: 'W84', reason: 'CloudWatch Logs are already encrypted by default.'
    }]);

    const api = new RestApi(this, 'M2C2Api', {
      defaultCorsPreflightOptions: {
        allowOrigins: ['*'],
        allowHeaders: [
          'Authorization',
          'Content-Type',
          'X-Amz-Date',
          'X-Amz-Security-Token',
          'X-Api-Key'
        ],
        allowMethods: [
          'GET',
          'POST',
          'OPTIONS'
        ],
        statusCode: 200
      },
      deploy: true,
      deployOptions: {
        accessLogDestination: new LogGroupLogDestination(apiLogGroup),
        accessLogFormat: AccessLogFormat.jsonWithStandardFields(),
        loggingLevel: MethodLoggingLevel.INFO,
        stageName: 'prod',
        tracingEnabled: true
      },
      description: 'Machine to Cloud Connectivity Rest API',
      endpointTypes: [EndpointType.REGIONAL]
    });
    this.apiEndpoint = `https://${api.restApiId}.execute-api.${Aws.REGION}.amazonaws.com/prod`;
    this.apiId = api.restApiId;

    const requestValidator = new RequestValidator(this, 'ApiRequestValidator', {
      restApi: api,
      validateRequestParameters: true,
      validateRequestBody: true
    });

    addCfnSuppressRules(api.node.findChild('Deployment') as Deployment, [{
      id: 'W68', reason: 'The solution does not require the usage plan.'
    }]);
    addCfnSuppressRules(api.node.findChild('DeploymentStage.prod') as Stage, [{
      id: 'W64', reason: 'The solution does not require the usage plan.'
    }]);

    const getAllIntegration = new Integration({
      type: IntegrationType.AWS_PROXY,
      integrationHttpMethod: 'POST',
      options: {
        contentHandling: ContentHandling.CONVERT_TO_TEXT,
        integrationResponses: [{ statusCode: '200' }],
        passthroughBehavior: PassthroughBehavior.WHEN_NO_MATCH,
        requestParameters: { 'integration.request.querystring.nextToken': 'method.request.querystring.nextToken' }
      },
      uri: `arn:${Aws.PARTITION}:apigateway:${Aws.REGION}:lambda:path/2015-03-31/functions/${connectionBuilderLambdaFunctionArn}/invocations`
    });
    const getAllMethodOptions: MethodOptions = {
      authorizationType: AuthorizationType.IAM,
      methodResponses: [{
        statusCode: '200',
        responseModels: {
          'application/json': { modelId: 'Empty' }
        }
      }],
      requestParameters: { 'method.request.querystring.nextToken': false },
      requestValidator: requestValidator
    };

    /**
     * Connections API
     * GET /connections
     * POST /connections
     * GET /connections/{connectionName}
     */
    const connectionsResource = api.root.addResource('connections');
    connectionsResource.addMethod(
      'POST',
      new Integration({
        type: IntegrationType.AWS_PROXY,
        integrationHttpMethod: 'POST',
        options: {
          contentHandling: ContentHandling.CONVERT_TO_TEXT,
          integrationResponses: [{ statusCode: '200' }],
          passthroughBehavior: PassthroughBehavior.WHEN_NO_MATCH
        },
        uri: `arn:${Aws.PARTITION}:apigateway:${Aws.REGION}:lambda:path/2015-03-31/functions/${connectionBuilderLambdaFunctionArn}/invocations`
      }),
      {
        authorizationType: AuthorizationType.IAM,
        methodResponses: [{
          statusCode: '200',
          responseModels: {
            'application/json': { modelId: 'Empty' }
          }
        }],
        requestValidator: requestValidator
      }
    );
    connectionsResource.addMethod('GET', getAllIntegration, getAllMethodOptions);

    const connectionResource = connectionsResource.addResource('{connectionName}');
    connectionResource.addMethod(
      'GET',
      new Integration({
        type: IntegrationType.AWS_PROXY,
        integrationHttpMethod: 'POST',
        options: {
          contentHandling: ContentHandling.CONVERT_TO_TEXT,
          integrationResponses: [{ statusCode: '200' }],
          passthroughBehavior: PassthroughBehavior.WHEN_NO_MATCH,
          requestParameters: { 'integration.request.path.connectionName': 'method.request.path.connectionName' }
        },
        uri: `arn:${Aws.PARTITION}:apigateway:${Aws.REGION}:lambda:path/2015-03-31/functions/${connectionBuilderLambdaFunctionArn}/invocations`
      }),
      {
        authorizationType: AuthorizationType.IAM,
        methodResponses: [{
          statusCode: '200',
          responseModels: {
            'application/json': { modelId: 'Empty' }
          }
        }],
        requestParameters: { 'method.request.path.connectionName': true },
        requestValidator: requestValidator
      }
    );

    /**
     * Logs API
     * GET /logs
     * GET /logs/{connectionName}
     */
    const logsResource = api.root.addResource('logs');
    logsResource.addMethod('GET', getAllIntegration, getAllMethodOptions);
    logsResource.addResource('{connectionName}').addMethod(
      'GET',
      new Integration({
        type: IntegrationType.AWS_PROXY,
        integrationHttpMethod: 'POST',
        options: {
          contentHandling: ContentHandling.CONVERT_TO_TEXT,
          integrationResponses: [{ statusCode: '200' }],
          passthroughBehavior: PassthroughBehavior.WHEN_NO_MATCH,
          requestParameters: {
            'integration.request.path.connectionName': 'method.request.path.connectionName',
            'integration.request.querystring.nextToken': 'method.request.querystring.nextToken'
          }
        },
        uri: `arn:${Aws.PARTITION}:apigateway:${Aws.REGION}:lambda:path/2015-03-31/functions/${connectionBuilderLambdaFunctionArn}/invocations`
      }),
      {
        authorizationType: AuthorizationType.IAM,
        methodResponses: [{
          statusCode: '200',
          responseModels: {
            'application/json': { modelId: 'Empty' }
          }
        }],
        requestParameters: {
          'method.request.path.connectionName': true,
          'method.request.querystring.nextToken': false
        },
        requestValidator: requestValidator
      }
    );

    /**
     * Sitewise API
     * GET /sitewise/{serverName}
     */
     const sitewiseResource = api.root.addResource('sitewise');
     sitewiseResource.addResource('{serverName}').addMethod(
       'GET',
       new Integration({
         type: IntegrationType.AWS_PROXY,
         integrationHttpMethod: 'POST',
         options: {
           contentHandling: ContentHandling.CONVERT_TO_TEXT,
           integrationResponses: [{ statusCode: '200' }],
           passthroughBehavior: PassthroughBehavior.WHEN_NO_MATCH,
           requestParameters: {
             'integration.request.path.serverName': 'method.request.path.serverName'
           }
         },
         uri: `arn:${Aws.PARTITION}:apigateway:${Aws.REGION}:lambda:path/2015-03-31/functions/${connectionBuilderLambdaFunctionArn}/invocations`
       }),
       {
         authorizationType: AuthorizationType.IAM,
         methodResponses: [{
           statusCode: '200',
           responseModels: {
             'application/json': { modelId: 'Empty' }
           }
         }],
         requestParameters: {
           'method.request.path.serverName': true
         },
         requestValidator: requestValidator
       }
     );

    connectionBuilderLambdaFunction.addPermission('ApiLambdaInvokePermission', {
      action: 'lambda:InvokeFunction',
      principal: new ServicePrincipal('apigateway.amazonaws.com'),
      sourceArn: api.arnForExecuteApi()
    });
  }
}