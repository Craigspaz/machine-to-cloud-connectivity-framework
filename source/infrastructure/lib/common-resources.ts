// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, Construct, RemovalPolicy } from '@aws-cdk/core';
import { Effect, Policy, PolicyStatement, ServicePrincipal } from '@aws-cdk/aws-iam';
import { BlockPublicAccess, Bucket, BucketAccessControl, BucketEncryption, CfnBucket } from '@aws-cdk/aws-s3';

/**
 * Machine to Cloud Connectivity Framework Common Resources
 */
export class CommonResourcesConstruct extends Construct {
  // CloudWatch Logs policy
  private cloudWatchLogsPolicy: Policy;
  // S3 Logging bucket
  private s3LoggingBucket: Bucket;
  // S3 Resource Storage bucket
  private s3Bucket: Bucket;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    this.cloudWatchLogsPolicy = new Policy(this, 'CloudWatchLogsPolicy', {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [
            'logs:CreateLogGroup',
            'logs:CreateLogStream',
            'logs:PutLogEvents'
          ],
          resources: [
            `arn:${Aws.PARTITION}:logs:${Aws.REGION}:${Aws.ACCOUNT_ID}:log-group:/aws/lambda/*`
          ]
        })
      ]
    });

    this.s3LoggingBucket = new Bucket(this, 'LogBucket', {
      accessControl: BucketAccessControl.LOG_DELIVERY_WRITE,
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
      encryption: BucketEncryption.S3_MANAGED,
      removalPolicy: RemovalPolicy.RETAIN
    });
    const cfnBucket = this.s3LoggingBucket.node.defaultChild as CfnBucket;
    cfnBucket.addMetadata('cfn_nag', {
      rules_to_suppress: [
        { id: 'W35', reason: 'This bucket is to store S3 logs, so it does not require access logs.' },
        { id: 'W51', reason: 'This bucket is to store S3 logs, so it does not require S3 policy.' }
      ]
    });

    this.s3Bucket = new Bucket(this, 'M2C2Bucket', {
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
      encryption: BucketEncryption.S3_MANAGED,
      removalPolicy: RemovalPolicy.RETAIN,
      serverAccessLogsBucket: this.s3LoggingBucket,
      serverAccessLogsPrefix: 'm2c2/'
    });
    this.s3Bucket.addToResourcePolicy(
      new PolicyStatement({
        effect: Effect.ALLOW,
        principals: [ new ServicePrincipal('lambda.amazonaws.com') ],
        actions: [
          's3:GetObject',
          's3:PutObject'
        ],
        resources: [
          `${this.s3Bucket.bucketArn}/*`
        ]
      })
    );
  }

  /**
   * Get CloudWatch Logs policy.
   * @return {Policy} CloudWatch Logs policy
   */
  getCloudWatchLogsPolicy(): Policy {
    return this.cloudWatchLogsPolicy;
  }

  /**
   * Get S3 logging bucket.
   * @return {Bucket} Logging bucket
   */
  getS3LoggingBucket(): Bucket {
    return this.s3LoggingBucket;
  }

  /**
   * Get S3 M2C2 bucket.
   * @return {Bucket} M2C2 bucket
   */
  getS3Bucket(): Bucket {
    return this.s3Bucket;
  }

  /**
   * Get S3 M2C2 bucket name
   * @return {string} M2C2 bucket name
   */
  getS3BucketName(): string {
    return this.s3Bucket.bucketName;
  }
}