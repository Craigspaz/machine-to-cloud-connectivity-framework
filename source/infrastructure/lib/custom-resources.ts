// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import {
  Aws,
  CfnCondition,
  CfnCustomResource,
  CfnMapping,
  Construct,
  CustomResource,
  Duration
} from '@aws-cdk/core';
import { Policy, Role, ServicePrincipal } from '@aws-cdk/aws-iam';
import { CfnFunction, Code, Function, Runtime } from '@aws-cdk/aws-lambda';
import { Bucket } from '@aws-cdk/aws-s3';

/**
 * CustomResourcesConstruct props
 */
export interface CustomResourcesConstructProps {
  /**
   * Policy for CloudWatch Logs
   */
  readonly cloudWatchLogsPolicy: Policy;
  /**
   * Existing Greengrass group ID
   */
  readonly existingGreengrassGroup: string;
  /**
   * Existing Kinesis Data Stream name
   */
  readonly existingKinesisStream: string;
  /**
   * Condition for sending anonymous usage
   */
  readonly sendAnonymousUsageCondition: CfnCondition;
  /**
   * Mapping for the source code
   */
  readonly sourceCodeMapping: CfnMapping;
  /**
   * Mapping for the solution information
   */
  readonly solutionMapping: CfnMapping;
}

/**
 * Machine to Cloud Connectivity Framework Custom Resources
 */
export class CustomResourcesConstruct extends Construct {
  // Solution UUID
  private uuid: string;

  constructor(scope: Construct, id: string, props: CustomResourcesConstructProps) {
    super(scope, id);

    const helperFunctionRole = new Role(this, 'HelperFunctionRole', {
      assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
      path: '/'
    });
    helperFunctionRole.attachInlinePolicy(props.cloudWatchLogsPolicy);

    const helperFunction = new Function(this, 'HelperFunction', {
      description: 'Machine to Cloud Connectivity custom resource function',
      handler: 'm2c2_helper_custom_resource.handler',
      runtime: Runtime.PYTHON_3_8,
      code: Code.fromBucket(
        Bucket.fromBucketName(this, 'SourceCodeBucket', `${props.sourceCodeMapping.findInMap('General', 'S3Bucket')}-${Aws.REGION}`),
        `${props.sourceCodeMapping.findInMap('General', 'KeyPrefix')}/m2c2-helper.zip`
      ),
      timeout: Duration.seconds(240),
      role: helperFunctionRole,
      environment: {
        SOLUTION_ID: props.solutionMapping.findInMap('Parameters', 'Id'),
        SOLUTION_VERSION: props.solutionMapping.findInMap('Parameters', 'Version'),
      }
    });
    const cfnFunction = helperFunction.node.defaultChild as CfnFunction;
    cfnFunction.addMetadata('cfn_nag', {
      rules_to_suppress: [
        { id: 'W58', reason: 'Miss alarm, and the function does have permission to write CloudWatch Logs.' }
      ]
    });

    const customUuid = new CustomResource(this, 'UUID', {
      serviceToken: helperFunction.functionArn,
      properties: {
        Resource: 'CreateUUID'
      }
    });
    this.uuid = customUuid.getAtt('UUID').toString();

    const sendAnonymousMetrics = new CustomResource(this, 'SendAnonymousMetrics', {
      serviceToken: helperFunction.functionArn,
      properties: {
        Resource: 'SendAnonymousMetrics',
        SolutionId: props.solutionMapping.findInMap('Parameters', 'Id'),
        UUID: this.uuid,
        Version: props.solutionMapping.findInMap('Parameters', 'Version'),
        Region: Aws.REGION,
        ExistingGreengrassGroup: props.existingGreengrassGroup,
        ExistingKinesisStream: props.existingGreengrassGroup
      }
    });
    const cfnSendAnonymousMetrics = sendAnonymousMetrics.node.defaultChild as CfnCustomResource;
    cfnSendAnonymousMetrics.cfnOptions.condition = props.sendAnonymousUsageCondition;
  }

  /**
   * Get solution UUID.
   * @return {string} UUID
   */
  getUuid(): string {
    return this.uuid;
  }
}