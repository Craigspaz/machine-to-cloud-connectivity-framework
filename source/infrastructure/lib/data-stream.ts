// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Construct } from '@aws-cdk/core';
import { Bucket } from '@aws-cdk/aws-s3';
import { KinesisStreamsToKinesisFirehoseToS3 } from '@aws-solutions-constructs/aws-kinesisstreams-kinesisfirehose-s3';

/**
 * DataStreamConstruct props
 */
export interface DataStreamConstructProps {
  /**
   * Logging S3 Bucket
   */
  readonly s3LoggingBucket: Bucket;
}

/**
 * Machine to Cloud Connectivity Framework Kinesis Streams, Kinesis Firehose, and S3
 */
export class DataStreamConstruct extends Construct {
  // Kinesis Data Stream name
  private kinesisStreamName: string;
  // Data S3 Bucket name
  private s3BucketName: string;

  constructor(scope: Construct, id: string, props: DataStreamConstructProps) {
    super(scope, id);

    const kinesisStreamsToKinesisFirehoseToS3 = new KinesisStreamsToKinesisFirehoseToS3(this, 'DataStream', {
      existingLoggingBucketObj: props.s3LoggingBucket,
      bucketProps: {
        serverAccessLogsPrefix: 'm2c2data/'
      }
    });
    this.kinesisStreamName = kinesisStreamsToKinesisFirehoseToS3.kinesisStream.streamName;
    this.s3BucketName = kinesisStreamsToKinesisFirehoseToS3.s3Bucket!.bucketName;
  }

  /**
   * Get Kinesis Data Stream name.
   * @return {string} Kinesis Data Stream name
   */
  getKinesisStreamName(): string {
    return this.kinesisStreamName;
  }

  /**
   * Get data S3 Bucket name.
   * @return {string} S3 Bucket name
   */
  getS3BucketName(): string {
    return this.s3BucketName;
  }
}