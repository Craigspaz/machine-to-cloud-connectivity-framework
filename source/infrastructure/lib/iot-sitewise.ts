// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, CfnDeletionPolicy, Construct } from '@aws-cdk/core';
import { CfnGateway } from '@aws-cdk/aws-iotsitewise';

/**
 * IoTSitewiseConstructProps props
 * @interface IoTSitewiseConstructProps
 */
export interface IoTSitewiseConstructProps {
  // Greengrass group ID
  readonly greengrassGroupId: string;
}

/**
 * @class
 * Machine to Cloud Connectivity Framework IoT Sitewise Construct.
 * It creates a common CloudWatch Logs policy for Lambda functions, a logging S3 bucket, and M2C2 S3 bucket.
 */
export class IoTSitewiseConstruct extends Construct {
  // IoT Sitewise gateway
  public iotSitewiseGateway: CfnGateway;

  constructor(scope: Construct, id: string, props: IoTSitewiseConstructProps) {
    super(scope, id);

    this.iotSitewiseGateway = new CfnGateway(this, 'Gateway', {
      gatewayName: `${Aws.STACK_NAME}-sitewise-gateway`,
      gatewayPlatform: {
        greengrass: {
          groupArn: `arn:${Aws.PARTITION}:greengrass:${Aws.REGION}:${Aws.ACCOUNT_ID}:/greengrass/groups/${props.greengrassGroupId}`
        }
      }
    });
    this.iotSitewiseGateway.cfnOptions.deletionPolicy = CfnDeletionPolicy.DELETE;
  }
}