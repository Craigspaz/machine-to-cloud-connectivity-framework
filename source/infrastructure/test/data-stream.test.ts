// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import { Stack } from '@aws-cdk/core';
import { Bucket, CfnBucket } from '@aws-cdk/aws-s3';
import { DataStreamConstruct } from '../lib/data-stream';

test('M2C2 data stream test', () => {
  const stack = new Stack();
  const s3LoggingBucket = new Bucket(stack, 'TestLoggingBucket');
  (s3LoggingBucket.node.defaultChild as CfnBucket).overrideLogicalId('TestLoggingBucket');
  const dataStream = new DataStreamConstruct(stack, 'TestDataStream', { s3LoggingBucket });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  expect(dataStream.getKinesisStreamName()).toBeDefined();
  expect(dataStream.getS3BucketName()).toBeDefined();
  expect(stack).toHaveResourceLike('AWS::S3::Bucket', {
    LoggingConfiguration: {
      DestinationBucketName: {
        Ref: 'TestLoggingBucket'
      },
      LogFilePrefix: 'm2c2data/'
    }
  });
});