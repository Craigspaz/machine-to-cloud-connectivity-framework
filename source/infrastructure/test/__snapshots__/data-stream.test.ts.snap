// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`M2C2 data stream test 1`] = `
Object {
  "Resources": Object {
    "TestDataStreamKinesisFirehoseToS3KinesisFirehose43AA011C": Object {
      "Properties": Object {
        "DeliveryStreamType": "KinesisStreamAsSource",
        "ExtendedS3DestinationConfiguration": Object {
          "BucketARN": Object {
            "Fn::GetAtt": Array [
              "TestDataStreamKinesisFirehoseToS3S3BucketAD612925",
              "Arn",
            ],
          },
          "BufferingHints": Object {
            "IntervalInSeconds": 300,
            "SizeInMBs": 5,
          },
          "CloudWatchLoggingOptions": Object {
            "Enabled": true,
            "LogGroupName": Object {
              "Ref": "TestDataStreamKinesisFirehoseToS3firehoseloggroup956F71B5",
            },
            "LogStreamName": Object {
              "Ref": "TestDataStreamKinesisFirehoseToS3firehoseloggroupfirehoselogstreamAD688049",
            },
          },
          "CompressionFormat": "GZIP",
          "EncryptionConfiguration": Object {
            "KMSEncryptionConfig": Object {
              "AWSKMSKeyARN": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":kms:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":alias/aws/s3",
                  ],
                ],
              },
            },
          },
          "RoleARN": Object {
            "Fn::GetAtt": Array [
              "TestDataStreamKinesisFirehoseToS3KinesisFirehoseRole22FE2669",
              "Arn",
            ],
          },
        },
        "KinesisStreamSourceConfiguration": Object {
          "KinesisStreamARN": Object {
            "Fn::GetAtt": Array [
              "TestDataStreamKinesisStream32860A4C",
              "Arn",
            ],
          },
          "RoleARN": Object {
            "Fn::GetAtt": Array [
              "TestDataStreamKinesisStreamsRole4C149BFE",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::KinesisFirehose::DeliveryStream",
    },
    "TestDataStreamKinesisFirehoseToS3KinesisFirehosePolicy2F5567C3": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:AbortMultipartUpload",
                "s3:GetBucketLocation",
                "s3:GetObject",
                "s3:ListBucket",
                "s3:ListBucketMultipartUploads",
                "s3:PutObject",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "TestDataStreamKinesisFirehoseToS3S3BucketAD612925",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "TestDataStreamKinesisFirehoseToS3S3BucketAD612925",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "logs:PutLogEvents",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":logs:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":log-group:",
                    Object {
                      "Ref": "TestDataStreamKinesisFirehoseToS3firehoseloggroup956F71B5",
                    },
                    ":log-stream:",
                    Object {
                      "Ref": "TestDataStreamKinesisFirehoseToS3firehoseloggroupfirehoselogstreamAD688049",
                    },
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestDataStreamKinesisFirehoseToS3KinesisFirehosePolicy2F5567C3",
        "Roles": Array [
          Object {
            "Ref": "TestDataStreamKinesisFirehoseToS3KinesisFirehoseRole22FE2669",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestDataStreamKinesisFirehoseToS3KinesisFirehoseRole22FE2669": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "firehose.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "TestDataStreamKinesisFirehoseToS3S3BucketAD612925": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "LifecycleConfiguration": Object {
          "Rules": Array [
            Object {
              "NoncurrentVersionTransitions": Array [
                Object {
                  "StorageClass": "GLACIER",
                  "TransitionInDays": 90,
                },
              ],
              "Status": "Enabled",
            },
          ],
        },
        "LoggingConfiguration": Object {
          "DestinationBucketName": Object {
            "Ref": "TestLoggingBucket",
          },
          "LogFilePrefix": "m2c2data/",
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": Object {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "TestDataStreamKinesisFirehoseToS3S3BucketPolicyA679D2DB": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "TestDataStreamKinesisFirehoseToS3S3BucketAD612925",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": "*",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "TestDataStreamKinesisFirehoseToS3S3BucketAD612925",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
              "Sid": "HttpsOnly",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "TestDataStreamKinesisFirehoseToS3firehoseloggroup956F71B5": Object {
      "DeletionPolicy": "Retain",
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W86",
              "reason": "Retention period for CloudWatchLogs LogGroups are set to 'Never Expire' to preserve customer data indefinitely",
            },
            Object {
              "id": "W84",
              "reason": "By default CloudWatchLogs LogGroups data is encrypted using the CloudWatch server-side encryption keys (AWS Managed Keys)",
            },
          ],
        },
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "TestDataStreamKinesisFirehoseToS3firehoseloggroupfirehoselogstreamAD688049": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "LogGroupName": Object {
          "Ref": "TestDataStreamKinesisFirehoseToS3firehoseloggroup956F71B5",
        },
      },
      "Type": "AWS::Logs::LogStream",
      "UpdateReplacePolicy": "Retain",
    },
    "TestDataStreamKinesisStream32860A4C": Object {
      "Properties": Object {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": Object {
          "EncryptionType": "KMS",
          "KeyId": "alias/aws/kinesis",
        },
      },
      "Type": "AWS::Kinesis::Stream",
    },
    "TestDataStreamKinesisStreamGetRecordsIteratorAgeAlarm8B028F82": Object {
      "Properties": Object {
        "AlarmDescription": "Consumer Record Processing Falling Behind, there is risk for data loss due to record expiration.",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "EvaluationPeriods": 1,
        "MetricName": "GetRecords.IteratorAgeMilliseconds",
        "Namespace": "AWS/Kinesis",
        "Period": 300,
        "Statistic": "Maximum",
        "Threshold": 2592000,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestDataStreamKinesisStreamReadProvisionedThroughputExceededAlarm72859B4F": Object {
      "Properties": Object {
        "AlarmDescription": "Consumer Application is Reading at a Slower Rate Than Expected.",
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "MetricName": "ReadProvisionedThroughputExceeded",
        "Namespace": "AWS/Kinesis",
        "Period": 300,
        "Statistic": "Average",
        "Threshold": 0,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestDataStreamKinesisStreamsRole4C149BFE": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "firehose.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "kinesis:DescribeStream",
                    "kinesis:GetShardIterator",
                    "kinesis:GetRecords",
                    "kinesis:ListShards",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::GetAtt": Array [
                      "TestDataStreamKinesisStream32860A4C",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "KinesisStreamsRoleRolePolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestLoggingBucket": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AccessControl": "LogDeliveryWrite",
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
